### PR TITLE
Uniformize Wallet API's path

### PR DESCRIPTION
**Pull request by NicolasDorier**
# Breaking changes

Many rename to uniformize everything by renaming everything on-chain from `/api/v1/stores/{storeId}/payment-methods/onchain/{cryptoCode}` to `/api/v1/stores/{storeId}/payment-methods/{paymentMethodId}/wallet`.

Also uniformizing `/generate`, `/preview` to have the same format. @ndeet 

This also remove the `Store payment methods (On-Chain)`, and put what was there in `Store Wallet (On-Chain)`.

Details:
* `/api/v1/stores/{storeId}/payment-methods/{paymentMethodId}/generate` => `/api/v1/stores/{storeId}/payment-methods/{paymentMethodId}/wallet/generate`
* `/api/v1/stores/{storeId}/payment-methods/{paymentMethodId}/preview` => `/api/v1/stores/{storeId}/payment-methods/{paymentMethodId}/wallet/preview`
* `/api/v1/stores/{storeId}/payment-methods/onchain/{cryptoCode}/wallet` => `/api/v1/stores/{storeId}/payment-methods/{paymentMethodId}/wallet`
* `/api/v1/stores/{storeId}/payment-methods/onchain/{cryptoCode}/wallet/feerate` => `/api/v1/stores/{storeId}/payment-methods/{paymentMethodId}/wallet/feerate`
* `/api/v1/stores/{storeId}/payment-methods/onchain/{cryptoCode}/wallet/address` => `/api/v1/stores/{storeId}/payment-methods/{paymentMethodId}/wallet/address`
* `/api/v1/stores/{storeId}/payment-methods/onchain/{cryptoCode}/wallet/transactions` => `/api/v1/stores/{storeId}/payment-methods/{paymentMethodId}/wallet/transactions`
* `/api/v1/stores/{storeId}/payment-methods/onchain/{cryptoCode}/wallet/transactions/{transactionId}` => 
 `/api/v1/stores/{storeId}/payment-methods/{paymentMethodId}/onchain/wallet/transactions/{transactionId}`
* `/api/v1/stores/{storeId}/payment-methods/onchain/{cryptoCode}/wallet/utxos` => `/api/v1/stores/{storeId}/payment-methods/{paymentMethodId}/wallet/utxos`
* `/api/v1/stores/{storeId}/payment-methods/onchain/{cryptoCode}/wallet/objects` => `/api/v1/stores/{storeId}/payment-methods/{paymentMethodId}/wallet/objects`
* `/api/v1/stores/{storeId}/payment-methods/onchain/{cryptoCode}/wallet/objects/{objectType}/{objectId}` => `/api/v1/stores/{storeId}/payment-methods/{paymentMethodId}/wallet/objects/{objectType}/{objectId}`
* `/api/v1/stores/{storeId}/payment-methods/onchain/{cryptoCode}/wallet/objects/{objectType}/{objectId}/links` => `/api/v1/stores/{storeId}/payment-methods/{paymentMethodId}/wallet/objects/{objectType}/{objectId}/links`
* `/api/v1/stores/{storeId}/payment-methods/onchain/{cryptoCode}/wallet/objects/{objectType}/{objectId}/links/{linkType}/{linkId}` => `/api/v1/stores/{storeId}/payment-methods/{paymentMethodId}/wallet/objects/{objectType}/{objectId}/links/{linkType}/{linkId}`

### CODE REVIEW

**Comment by ndeet** [comment]
On payment methods (pm) we only need the generic one which is neither onchain nor LN imo, so this can go

> Many rename to uniformize everything by renaming everything on-chain from /api/v1/stores/{storeId}/payment-methods/onchain/{crypto[Code](https://github.com/btcpayserver/btcpayserver)} to /api/v1/stores/{storeId}/payment-methods/{paymentMethodId}/onchain.

As discussed I would suggest to use distinct path for wallet related endpoints. Instead of `paymentmethods` use `wallets` , don't use `onchain` at all?

Examples:
/api/v1/stores/{storeId}/payment-methods/{paymentMethodId}/generate => /api/v1/stores/{storeId}/wallets/{paymentMethodId}/generate
...

/api/v1/stores/{storeId}/payment-methods/onchain/{cryptoCode}/wallet => /api/v1/stores/{storeId}/wallets/{paymentMethodId}

/api/v1/stores/{storeId}/payment-methods/onchain/{cryptoCode}/wallets/feerate => /api/v1/stores/{storeId}/wallets/{paymentMethodId}/feerate

/api/v1/stores/{storeId}/payment-methods/onchain/{cryptoCode}/wallet/address => /api/v1/stores/{storeId}/wallets/{paymentMethodId}/address

/api/v1/stores/{storeId}/payment-methods/onchain/{cryptoCode}/wallet/transactions => /api/v1/stores/{storeId}/wallets/{paymentMethodId}/transactions
...

/api/v1/stores/{storeId}/payment-methods/onchain/{cryptoCode}/wallet/objects => /api/v1/stores/{storeId}/wallets/{paymentMethodId}/objects

/api/v1/stores/{storeId}/payment-methods/onchain/{cryptoCode}/wallet/objects/{objectType}/{objectId} => /api/v1/stores/{storeId}/wallets/{paymentMethodId}/objects/{objectType}/{objectId}
...

**Comment by NicolasDorier** [comment]
`/api/v1/stores/{storeId}/wallets/{paymentMethodId}/generate`

What if another payment method come in with their own wallet with different request/response? Then it will need to become something like

`/api/v1/stores/{storeId}/wallets-xmr/{paymentMethodId}/generate` ?

Because if that's the case, why couldn't they have the same prefix? It's a nice way to see that all payment methods share a common path.

Also, there isn't any wallet concept in BTCPay that is independant from payment methods, so I don't think it's a good idea to expose it as a new kind of object.

**Comment by ndeet** [comment]
> `/api/v1/stores/{storeId}/wallets/{paymentMethodId}/generate`
> 
> What if another payment method come in with their own wallet with different request/response? Then it will need to become something like
> 
> `/api/v1/stores/{storeId}/wallets-xmr/{paymentMethodId}/generate` ?
> 
> Because if that's the case, why couldn't they have the same prefix? It's a nice way to see that all payment methods share a common path.

Hmm good point but not sure we should care for altcoin edgecases. If they want their own endpoints they can do it on ../store/{storeId}/plugins/{pluginId}/CUSTOM/SOMETHING

> Also, there isn't any wallet concept in BTCPay that is independant from payment methods, so I don't think it's a good idea to expose it as a new kind of object.

Ok did not know but its a logical diff maybe? Or instead of `wallets` maybe `payment-method/pmId/wallet
`

> Last but not the least. I could add a one line trick so people with API using the old path doesn't break!

That would be really nice and deprecate with some next major verion

File: BTCPayServer/Hosting/Startup.cs [hunk 1]
```diff
@@ -281,6 +281,10 @@ private void ConfigureCore(IApplicationBuilder app, IWebHostEnvironment env, ISe
             var rewriteOptions = new RewriteOptions();
             rewriteOptions.AddRewrite("_blazor/(negotiate|initializers|disconnect)$", "/_blazor/$1", skipRemainingRules: true);
             rewriteOptions.AddRewrite("_blazor$", "/_blazor", skipRemainingRules: true);
+
+            // A rewrite rule to support the old API
+            rewriteOptions.AddRewrite("api/v1/stores/([^/]+)/payment-methods/[Oo]n[Cc]hain/([^/]+)/(preview|generate)", "/api/v1/stores/$1/payment-methods/$2-CHAIN/wallet/$3", skipRemainingRules: true);
```
Comment (ndeet) [inline BTCPayServer/Hosting/Startup.cs:286]: nice

**Comment by ndeet** [comment]
looks much cleaner now, imo :+1:

File: BTCPayServer.Client/BTCPayServerClient.OnChainWallet.Objects.cs [hunk 2]
```diff
@@ -15,7 +15,7 @@ public virtual async Task<OnChainWalletObjectData> GetOnChainWalletObject(string
             parameters.Add("includeNeighbourData", v);
         try
         {
-            return await SendHttpRequest<OnChainWalletObjectData>($"api/v1/stores/{storeId}/payment-methods/onchain/{cryptoCode}/wallet/objects/{objectId.Type}/{objectId.Id}", parameters, HttpMethod.Get, token);
+            return await SendHttpRequest<OnChainWalletObjectData>($"api/v1/stores/{storeId}/payment-methods/{cryptoCode}-CHAIN/wallet/objects/{objectId.Type}/{objectId.Id}", parameters, HttpMethod.Get, token);
```
[the remark this answers is not in the document]
Comment (NicolasDorier) [reply BTCPayServer.Client/BTCPayServerClient.OnChainWallet.Objects.cs:18]: yeah... I think I should just update the client library in C# to align to it... but this bring a bunch of changes everywhere.

**Pull request merged**